### ISSUE TITLE
AttributeError occurs when do jobs with MULTIPLE_FORCE_EVALS

### ISSUE THREAD

**Issue by xjf729** [aiidateam/aiida-cp2k#20]
### My input parameter
When I use aiida-cp2k plugin to do jobs whose `.inp` file should include more than one `FORCE_EVAL`(for example, redox potential job), `AttributeError` will occur.

There is my main input parameters:

```python
input_para = {
    'GLOBAL': global_set,
    'MULTIPLE_FORCE_EVALS': multiple_force_evals,
    'FORCE_EVAL': [force_eval_mixed, force_eval_red, force_eval_ox],
    'MOTION': motion
    }
```

in which, `force_eval_mixed`, `force_eval_red` and `force_eval_ox` are all `dict` type.
### Error logs
By `verdi calculation logshow `, I got these error logs:

```
*** 133: SUBMISSIONFAILED
*** Scheduler output: N/A
*** Scheduler errors: N/A
*** 1 LOG MESSAGES:
+-> ERROR at 2019-04-09 16:35:08.923204+00:00
|   Submission of calc 133 failed, check also the log file! Traceback: Traceback (most recent call last):
|     File "/home/aiida/.local/lib/python2.7/site-packages/aiida/daemon/execmanager.py", line 475, in submit_calc
|       folder, use_unstored_links=False)
|     File "/home/aiida/.local/lib/python2.7/site-packages/aiida/orm/implementation/general/calculation/job/__init__.py", line 1472, in _presubmit
|       calcinfo = self._prepare_for_submission(folder, inputdict)
|     File "/home/aiida/.local/lib/python2.7/site-packages/aiida_cp2k/calculations/__init__.py", line 115, in _prepare_for_submission
|       inp.add_keyword('FORCE_EVAL/SUBSYS/CELL/'+a, val)
|     File "/home/aiida/.local/lib/python2.7/site-packages/aiida_cp2k/calculations/__init__.py", line 222, in add_keyword
|       self._add_keyword_low(kwpath.split("/"), value, self.params)
|     File "/home/aiida/.local/lib/python2.7/site-packages/aiida_cp2k/calculations/__init__.py", line 233, in _add_keyword_low
|       self._add_keyword_low(kwpath[1:], value, params[kwpath[0]])
|     File "/home/aiida/.local/lib/python2.7/site-packages/aiida_cp2k/calculations/__init__.py", line 228, in _add_keyword_low
|       elif kwpath[0] not in params.keys():
|   AttributeError: 'list' object has no attribute 'keys'
```

### Possible reason
https://github.com/aiidateam/aiida-cp2k/blob/aa962008164bc5cec42be874f8f9e24498076b69/aiida_cp2k/calculations/__init__.py#L115

You use `FORCE_EVAL/SUBSYS/CELL/` directly in this line, which means I cannot assign `list` type to `key` `FORCE_EVAL` in my input parameters. So I can just work with single `FORCE_EVAL`. 

I wonder know if there is a way to give input parameters which can do jobs with `MULTIPLE_FORCE_EVALS`.

**Comment by edditler** [comment]
As a pointer in case no one else can help you. What you want to do is achieved in another app: https://github.com/cpignedoli/aiidalab-empa-surfaces/blob/master/slab/slabwork.py#L210-L217

I can't tell you if those workflows are also broken because of a change in the plugin. On first sight, they also define inp['FORCE_EVAL'] as a list of dictionaries.

**Comment by yakutovicha** [comment]
Hi @xjf729 ,

Can you maybe send the entire input script that you want to run. Without seeing it, it is hard to judge whether the problem is in the plugin or in your input.

In case the problem is in the plugin - I am happy to modify it to fix the issue.

**Comment by xjf729** [comment]
Hi @yakutovicha 
This is my input script. The main part is from line 191.

[https://github.com/xjf729/aiida_cp2k_test/blob/209c7ca77d66c618cf3217e5e0b93714cc2cfe4f/redox_potential.py#L191-L223](url)

I will appreciate it if you could tell me the reason which raises the  AttributeError.

**Comment by dev-zero** [comment]
@xjf729 I haven't figured out the problem yet, but are you aware that a `dict.update` does not do a recursive update? AFAIK https://github.com/xjf729/aiida_cp2k_test/blob/996b9d83759b0297d656f0309c023241f28a29e7/redox_potential.py#L116 will replace the whole `SCF` subsection with that print subsection.

A simpler example:

```terminal
In [4]: d1={'foo': {'bar': {'something': 'here'}}}
In [5]: d1.update({'foo': {'bar': {'anotherthing': 'here'}}})
In [6]: d1
Out[6]: {'foo': {'bar': {'anotherthing': 'here'}}}
```

**Comment by dev-zero** [comment]
@xjf729 can you please report also the version number/source for your installation of the aiida-cp2k plugin?
Based on the line numbers from your traceback I'd guess that it is a rather old version where the functionality to add multiple sections for the same keyword was limited to the `KINDS` section.

**Comment by xjf729** [comment]
Hi @dev-zero , 
Thank you for pointing out my mistake, I had ignored the problem with unexpected results of `dict.update`.

Anyway, I used aiida-cp2k version 0.10 (matches aiida_core version 0.12.3), and I think the problem is in the `inp.add_keyword('FORCE_EVAL/SUBSYS/CELL/'+a, val)` where `FORCE_EVAL/SUBSYS/CELL/` was directly used. Because when I simply change [line110-119 in aiida_cp2k/calculations/__init__.py](https://github.com/aiidateam/aiida-cp2k/blob/v0.10/aiida_cp2k/calculations/__init__.py#L110-L119) to the following code, I can run well without any error logs.
```python
        if structure is not None:
            struct_fn = tempfolder.get_abs_path(self._COORDS_FILE_NAME)
            structure.export(struct_fn, fileformat="xyz")
            if isinstance(inp.params['FORCE_EVAL'], dict):
                for i, a in enumerate('ABC'):
                    val = '{:<15} {:<15} {:<15}'.format(*structure.cell[i])
                    inp.add_keyword('FORCE_EVAL/SUBSYS/CELL/'+a, val)
                topo = "FORCE_EVAL/SUBSYS/TOPOLOGY"
                inp.add_keyword(topo + "/COORD_FILE_NAME", self._COORDS_FILE_NAME)
                inp.add_keyword(topo + "/COORD_FILE_FORMAT", "XYZ")
            elif isinstance(inp.params['FORCE_EVAL'], list):
                for num_eval, force_eval in enumerate(inp.params['FORCE_EVAL']):
                    inp_force_eval = Cp2kInput(force_eval)
                    for i, a in enumerate('ABC'):
                        val = '{:<15} {:<15} {:<15}'.format(*structure.cell[i])
                        inp_force_eval.add_keyword('SUBSYS/CELL/'+a, val)
                    topo = "SUBSYS/TOPOLOGY"
                    inp_force_eval.add_keyword(topo + "/COORD_FILE_NAME", self._COORDS_FILE_NAME)
                    inp_force_eval.add_keyword(topo + "/COORD_FILE_FORMAT", "XYZ")
                    inp.params['FORCE_EVAL'][num_eval].update(inp_force_eval.params)
            else:
                raise InputValidationError("invalide FORCE_EVAL")
```

**Comment by dev-zero** [comment]
ugh, yeah, the `Cp2kInit.add_keyword()` doesn't know anything about that expansion of lists to repeated sections...
Please be aware that in your version `.add_keyword()` may still modify the input (`force_eval`) in your case. See https://github.com/aiidateam/aiida-cp2k/pull/22/commits/bf6fd6fc6e644ac930aedbc118d5c797ffe83aa5

**Comment by yakutovicha** [comment]
Fixed in 31bbfa0371ebf617b7e7a1f9bcd7820e9514860a, will go into 1.0.0b3 release
